plugin/sampling/strategystore/adaptive/processor.go

```diff
@@ -517,3 +518,41 @@ func (p *processor) generateDefaultSamplingStrategyResponse() *sampling.Sampling
 		},
 	}
 }
+
+type strategyStore struct {
+	logger           *zap.Logger
+	storedStrategies atomic.Value // holds *storedStrategies
+	ctx              context.Context
+	cancelFunc       context.CancelFunc
+}
+
+type storedStrategies struct {
+	defaultStrategy   *sampling.SamplingStrategyResponse
+	serviceStrategies map[string]*sampling.SamplingStrategyResponse
+}
+
+// GetSamplingStrategy implements StrategyStore#GetSamplingStrategy.
+func (h *strategyStore) GetSamplingStrategy(_ context.Context, serviceName string) (*sampling.SamplingStrategyResponse, error) {
+	ss := h.storedStrategies.Load().(*storedStrategies)
+	serviceStrategies := ss.serviceStrategies
+	if strategy, ok := serviceStrategies[serviceName]; ok {
+		return strategy, nil
+	}
+	h.logger.Debug("sampling strategy not found, using default", zap.String("service", serviceName))
+	return ss.defaultStrategy, nil
+}
+
+// NewStrategyStore creates a strategy store that holds adaptive sampling strategies.
+func NewStrategyStore(options Options, logger *zap.Logger) (ss.StrategyStore, error) {
+	ctx, cancelFunc := context.WithCancel(context.Background())
+	h := &strategyStore{
+		logger:     logger,
+		ctx:        ctx,
```

Comment:
If we can't capture the context, how will we cancel the child goroutines on an interrupt.

Reply:
Also, being new, I am interested to know why is it considered as anti-pattern.

Reply:
From the [golang docs on context](https://golang.org/pkg/context/):
> Do not store Contexts inside a struct type; instead, pass a Context explicitly to each function that needs it. The Context should be the first parameter, typically named ctx

Though they don't provide an explanation why, I believe this is an anti-pattern because context should be used on a per-request scope. Storing context in a struct means there's potential for race conditions on context being overridden in the struct and leading to unexpected behaviour. Imagine hundreds of web server requests being handled simultaneously, where trace information is stored in this same context object.

> If we can't capture the context, how will we cancel the child goroutines on an interrupt.

As long as the context is passed into the goroutine, calling the `cancelFunc` will allow the child goroutines to be early via `<-ctx.Done()`. I think this blog post provides a good example and explanation: https://levelup.gitconnected.com/how-to-use-context-to-manage-your-goroutines-like-a-boss-ef1e478919e6

Reply:
Thank you @albertteoh for explanation. That makes a lot of sense. I will read the links as well to get more details.

I have a followup question: does this means we are following this anti-pattern in static sampling already: https://github.com/jaegertracing/jaeger/blob/master/plugin/sampling/strategystore/static/strategy_store.go#L40-L47

Reply:
> does this means we are following this anti-pattern in static sampling already:

yes, because this call can simply pass the context instead of storing it

https://github.com/jaegertracing/jaeger/blob/bfd17cde88453736f61956cefe53a8482d45be11/plugin/sampling/strategystore/static/strategy_store.go#L79

Reply:
Fair question, yes, I think this is following that anti-pattern. If I'm not mistaken, that `ctx` object should be passed into the `h.autoUpdateStrategies` goroutine.

Reply:
PR to address it: #2823